src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RandomNumberGenerator.cs

```diff
@@ -165,6 +166,200 @@ public static byte[] GetBytes(int count)
+        /// </param>
+        /// <remarks>
+        ///   The behavior of this is the same as using <seealso cref="GetItems{T}(ReadOnlySpan{T}, Span{T})" /> and
+        ///   specifying hexadecimal characters as the choices. This implementation is optimized specifically for
+        ///   hexadecimal characters.
+        /// </remarks>
+        public static void GetHexString(Span<char> destination, bool lowercase = false)
+        {
+            if (destination.IsEmpty)
+                return;
+
+            GetHexStringCore(destination, lowercase);
+        }
+
+        /// <summary>
+        ///   Creates a string filled with cryptographically random hexadecimal characters.
+        /// </summary>
+        /// <param name="stringLength">The length of string to create.</param>
+        /// <param name="lowercase">
+        ///   <see langword="true" /> if the hexadecimal characters should be uppercase; <see langword="true" /> if lowercase.
+        ///   The default is <see langword="false" />.
+        /// </param>
+        /// <returns>A string populated with random hexadecimal characters.</returns>
+        /// <remarks>
+        ///   The behavior of this is the same as using <seealso cref="GetString" /> and
+        ///   specifying hexadecimal characters as the choices. This implementation is optimized specifically for
+        ///   hexadecimal characters.
+        /// </remarks>
+        public static string GetHexString(int stringLength, bool lowercase = false)
+        {
+            if (stringLength == 0)
+                return string.Empty;
+
+            return string.Create(stringLength, lowercase, GetHexStringCore);
+        }
+
+        /// <summary>
+        ///   Performs an in-place shuffle of a span using cryptographically random number generation.
+        /// </summary>
+        /// <param name="values">The span to shuffle.</param>
+        /// <typeparam name="T">The type of span.</typeparam>
+        public static void Shuffle<T>(Span<T> values)
+        {
+            int n = values.Length;
+
+            for (int i = 0; i < n - 1; i++)
+            {
+                int j = GetInt32(i, n);
+
+                if (i != j)
+                {
+                    T temp = values[i];
+                    values[i] = values[j];
+                    values[j] = temp;
+                }
+            }
+        }
+
+        private static void GetHexStringCore(Span<char> destination, bool lowercase)
+        {
+            Debug.Assert(!destination.IsEmpty);
+
+            const int RandomBufferSize = 64; // If this changes, the tests need to be updated since they try to exercise boundary conditions.
+            Span<byte> randomBuffer = stackalloc byte[RandomBufferSize];
+            HexConverter.Casing casing = lowercase ? HexConverter.Casing.Lower : HexConverter.Casing.Upper;
+
+            // Don't overfill the buffer if the destination is smaller than the buffer size. We need to round up when
```

Comment:
what's the concern with overfilling? wasting entropy? is it measurably more expensive? etc

Reply:
> is it measurably more expensive?

On non-desktop platforms like a budget Android device, yeah, it's noticeable. On desktop systems it's negligible.